src/memory/db_stores/redis_store.rs

```diff
@@ -0,0 +1,233 @@
+    /// Connects to a Redis server using the given URL.
+    pub async fn connect(url: &str) -> Result<Self, RedisStoreError> {
+        let client = redis::Client::open(url)?;
+        let mut manager = client.get_connection_manager().await?;
+        let write_script_hash = redis::cmd("SCRIPT")
+            .arg("LOAD")
+            .arg(GUARDED_WRITE_LUA_SCRIPT)
+            .query_async(&mut manager)
+            .await?;
+        Ok(Self {
+            manager,
+            write_script_hash,
+            _marker_adr: PhantomData,
+        })
+    }
+
+    /// Connects to a Redis server with a `ConnectionManager`.
+    pub async fn connect_with_manager(manager: ConnectionManager) -> Result<Self, RedisStoreError> {
+        Ok(Self {
+            manager: manager.clone(),
+            write_script_hash: redis::cmd("SCRIPT")
+                .arg("LOAD")
+                .arg(GUARDED_WRITE_LUA_SCRIPT)
+                .query_async(&mut manager.clone())
+                .await?,
+            _marker_adr: PhantomData,
+        })
+    }
+
+    pub async fn clear_indexes(&self) -> Result<(), redis::RedisError> {
+        redis::cmd("FLUSHDB")
+            .query_async(&mut self.manager.clone())
+            .await
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RedisStoreError(String);
+
+impl std::error::Error for RedisStoreError {}
+
+impl From<redis::RedisError> for RedisStoreError {
+    fn from(err: redis::RedisError) -> Self {
+        Self(err.to_string())
+    }
+}
+
+impl Display for RedisStoreError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Redis Memory Error: {}", self.0)
+    }
+}
+
+impl<
+    Address: Send + Sync + Hash + Eq + Debug + Clone + Deref<Target = [u8; ADDRESS_LENGTH]>,
+    const ADDRESS_LENGTH: usize,
+    const WORD_LENGTH: usize,
+> MemoryADT for RedisStore<Address, WORD_LENGTH>
+{
+    type Address = Address;
+    type Error = RedisStoreError;
+    type Word = [u8; WORD_LENGTH];
+
+    async fn batch_read(
+        &self,
+        addresses: Vec<Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        let refs: Vec<&[u8; ADDRESS_LENGTH]> =
+            addresses.iter().map(|address| address.deref()).collect();
+        self.manager
+            .clone()
+            .mget::<_, Vec<_>>(&refs)
+            .await
+            .map_err(Self::Error::from)
+    }
+
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        bindings: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<Self::Word>, Self::Error> {
+        let (guard_address, guard_value) = guard;
+        let mut cmd = redis::cmd("EVALSHA")
+            .arg(self.write_script_hash.clone())
+            .arg(0)
+            .arg(&*guard_address)
+            .clone(); // Why cloning is necessary : https://stackoverflow.com/questions/64728534/how-to-resolve-creates-a-temporary-variable-which-is-freed-while-still-in-use
```

Comment:
I do not see any valid justification for cloning in this link. Please detail it here.